src/file_config.rs

```diff
@@ -0,0 +1,270 @@
+            info!("Configured {dup}source {id} from {}", can.display());
+            configs.insert(id.clone(), source.clone());
+
+            let path = match source {
+                FileConfigSrc::Obj(pmt) => pmt.path,
+                FileConfigSrc::Path(path) => path,
+            };
+            results.insert(id.clone(), create_source(id, path).await?);
+        }
+    }
+
+    if let Some(paths) = cfg.paths {
+        for path in paths {
+            let is_dir = path.is_dir();
+            let dir_files = if is_dir {
+                // directories will be kept in the config just in case there are new files
+                directories.push(path.clone());
+                path.read_dir()?
+                    .filter_map(std::result::Result::ok)
+                    .filter(|f| {
+                        f.path().extension().filter(|e| *e == extension).is_some()
+                            && f.path().is_file()
+                    })
+                    .map(|f| f.path())
+                    .collect()
+            } else if path.is_file() {
+                vec![path]
+            } else {
+                return Err(InvalidFilePath(path.canonicalize().unwrap_or(path)));
+            };
+            for path in dir_files {
+                let can = path.canonicalize()?;
+                if files.contains(&can) {
+                    if !is_dir {
+                        warn!("Ignoring duplicate MBTiles path: {}", can.display());
+                    }
+                    continue;
+                }
+                let id = path.file_stem().map_or_else(
+                    || "_unknown".to_string(),
+                    |s| s.to_string_lossy().to_string(),
+                );
+                let source = FileConfigSrc::Path(path);
+                let id = idr.resolve(&id, can.to_string_lossy().to_string());
+                info!("Configured source {id} from {}", can.display());
+                files.insert(can);
+                configs.insert(id.clone(), source.clone());
+
+                let path = match source {
+                    FileConfigSrc::Obj(pmt) => pmt.path,
+                    FileConfigSrc::Path(path) => path,
+                };
+                results.insert(id.clone(), create_source(id, path).await?);
+            }
+        }
+    }
+
+    *config = FileConfigEnum::Config(FileConfig {
+        paths: OneOrMany::new_opt(directories),
+        sources: Some(configs),
+        unrecognized: cfg.unrecognized,
+    });
+
+    Ok(results)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+    use std::path::PathBuf;
+
+    use indoc::indoc;
+
+    use crate::file_config::{FileConfigEnum, FileConfigSource, FileConfigSrc};
+
+    #[test]
+    fn parse() {
+        let cfg = serde_yaml::from_str::<FileConfigEnum>(indoc! {"
+            paths:
+              - /dir-path
+              - /path/to/pmtiles2.pmtiles
+            sources:
+                pm-src1: /tmp/pmtiles.pmtiles
```

Comment:
Probably we should also add the `*.mbtiles` test here?

Reply:
there is no point because this test is not specific to pmtiles - it decodes a sub-element under `pmtiles:` or `mbtiles:` keys.  I will change the filenames to make it clearer.